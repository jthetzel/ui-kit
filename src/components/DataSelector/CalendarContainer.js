import styled from 'react-emotion';

export default styled.div`
  position: absolute;
  top: 50px;
  left: 0;
  background-color: white;
  z-index: 5;
  
  input {
    display: block;
  }
  
  ${props => props.multiple && `
    .Selectable .DayPicker-Day--selected:not(.DayPicker-Day--start):not(.DayPicker-Day--end):not(.DayPicker-Day--outside) {
      background-color: #f0f8ff !important;
      color: #4a90e2;
    }
    .Selectable .DayPicker-Day {
      border-radius: 0 !important;
    }
    .Selectable .DayPicker-Day--start {
      border-top-left-radius: 50% !important;
      border-bottom-left-radius: 50% !important;
    }
    .Selectable .DayPicker-Day--end {
      border-top-right-radius: 50% !important;
      border-bottom-right-radius: 50% !important;
    }`};
`;
