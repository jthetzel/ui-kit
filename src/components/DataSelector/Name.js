import styled from 'react-emotion';

export default styled.div`
  font-family: 'Noto Sans', serif;
  font-size: 12px;
  color: #303d41;
`;
