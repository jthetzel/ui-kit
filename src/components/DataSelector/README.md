DataSelector

```js
<DataSelector onChange={(value) => console.log(value)} />
```

DataSelector multiple

```js
<DataSelector onChange={(value) => console.log(value)} multiple />
```