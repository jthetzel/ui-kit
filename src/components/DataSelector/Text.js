import styled from 'react-emotion';

export default styled.div`
  font-family: 'Noto Sans', serif;
  font-size: 16px;
`;
