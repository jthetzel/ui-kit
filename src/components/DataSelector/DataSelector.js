import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { DateUtils } from 'react-day-picker';
import enhanceWithClickOutside from 'react-click-outside';
import Calendar from '../Calendar';
import dateParse from '../../utils/dateParse';

// Style
import Container from './Container';
import CalendarContainer from './CalendarContainer';
import Text from './Text';
import Name from './Name';

/**
 * @component
 */

class DataSelector extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      isActive: this.props.isActive,
      from: this.props.from,
      to: this.props.to
    };

    this.toggle = this.toggle.bind(this);
    this.onChangeHour = this.onChangeHour.bind(this);
    this.handleDayClick = this.handleDayClick.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
    this._renderCalendar = this._renderCalendar.bind(this);
    this._renderDate = this._renderDate.bind(this);
  }

  toggle() {
    this.setState({ isActive: !this.state.isActive });
  }

  handleClickOutside() {
    this.setState({ isActive: false });
  }

  handleDayClick(day) {
    const { from } = this.state;
    const { multiple, onChange } = this.props;
    if (multiple) {
      const range = DateUtils.addDayToRange(day, this.state);
      onChange(range);
      this.setState(range);
    } else if (from.getHours() !== 12) {
      const newDate = new Date(day.setHours(from.getHours()));
      onChange({ from: newDate });
      this.setState({ from: newDate });
    } else {
      onChange({ from: day });
      this.setState({ from: day });
    }
  }

  onChangeHour(e) {
    const newDate = new Date(this.state.from.setHours(e.target.value));
    this.props.onChange({ from: newDate });
    this.setState({ from: newDate });
  }

  _renderDate() {
    const { from, to } = this.state;
    const start = dateParse(from);
    const end = dateParse(to);

    return to ? `${start} – ${end}` : start;
  }

  _renderCalendar() {
    const { from, to } = this.state;
    const { multiple } = this.props;
    const modifiers = { start: from, end: to };

    return (
      <CalendarContainer
        onClick={e => e.stopPropagation()}
        multiple={multiple}
      >
        <Calendar
          className={multiple ? 'Selectable' : null}
          modifiers={multiple ? modifiers : null}
          selectedDays={[from, { from, to }]}
          onDayClick={this.handleDayClick}
        />
        {
          multiple &&
            <input
              key="slider"
              type="range"
              max={23}
              onChange={this.onChangeHour}
            />
        }
      </CalendarContainer>
    );
  }

  render() {
    const { isActive } = this.state;

    return (
      <Container onClick={this.toggle}>
        <Name>{this.props.name}</Name>
        <Text>{this._renderDate()}</Text>
        { isActive && this._renderCalendar() }
      </Container>
    );
  }
}

DataSelector.propTypes = {
  isActive: PropTypes.bool,
  from: PropTypes.instanceOf(Date),
  to: PropTypes.node,
  multiple: PropTypes.bool,
  name: PropTypes.node,
  onChange: PropTypes.func.isRequired
};

DataSelector.defaultProps = {
  isActive: false,
  from: new Date(),
  to: PropTypes.Null,
  multiple: false,
  name: 'Дата'
};

export default enhanceWithClickOutside(DataSelector);
