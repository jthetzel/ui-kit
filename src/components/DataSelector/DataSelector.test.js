import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import DataSelector from '../DataSelector';

test('DataSelector does not crash', () => {
  const tree = shallow(<DataSelector onChange={() => {}} />);
  expect(toJson(tree)).toMatchSnapshot();
});
