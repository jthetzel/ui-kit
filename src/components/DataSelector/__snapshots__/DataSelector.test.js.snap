// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSelector does not crash 1`] = `
<DataSelector
  from={2017-12-14T09:49:06.149Z}
  isActive={false}
  multiple={false}
  name="Дата"
  onChange={[Function]}
/>
`;
